### DISCUSSION TITLE
Storing Graphs in Postgres

### DISCUSSION THREAD

**Discussion by MikeB2019x** [networkx/networkx#5253] [Q&A] [answered]
I would like to store graph structure in a way that is easily accessible to other users.  Currently I save graphs in .graphml format on a commonly accessible drive.  This is not always suitable and results in very crude name versioning as changes are made to the graph by users.  I thought perhaps I could save the graph in a database.  I am familiar with Neo4J and other commercial products but only have access to a postgres database.  If it's possible, is there a recommended way of storing multiple graphs in postgres?  Below is some code that creates a graph and converts it to a table that I could then ingest in to postgres.  It's very crude and inefficient.  I would see the end result as broken in to three tables in postgres: graph meta data, node data, link data.

```
import networkx as nx
import pandas as pd
from random import randint


# Create a directed graph
G = nx.circular_ladder_graph(5).to_directed()
G.name = 'test'
# Randomize edge weights
nx.set_edge_attributes(G, {e: {'weight': randint(1, 9)} for e in G.edges})

data = nx.node_link_data(G)

test = pd.DataFrame.from_dict(data, orient='index').transpose()

pd.json_normalize(test.nodes[0]).merge(pd.json_normalize(
    test.links[0]), left_index=True, right_index=True, how='right').merge(
    test[['directed', 'multigraph', 'graph']], left_index=True, 
    right_index=True, how='left')[['directed', 'multigraph', 
                                           'graph', 'id', 'weight', 'source', 'target']]
```

**Comment by rossbar** [comment]
>  If it's possible, is there a recommended way of storing multiple graphs in postgres?

The short answer, I would think, is no - there is no **recommended** way. The NetworkX graph data structure is based on a nested-dicts-like data structure of various levels depending on e.g. whether the graph is a multigraph or not. The advantage of this data structure is that it is very flexible: you can have arbitrary objects as node or edge attributes; there is no requirement that all nodes or edges have the same attributes; and there is no "wasted" data to indicate e.g. the absence of an edge.

I'm no database expert, but I don't think that dict-of-dicts maps very well to relational databases. For graphs that leverage the full flexibility of the data structure, I would imagine that non-relational databases generally tend to be a better fit. However, I'm sure there are cases where some graphs can be in a relational database in a very straightforward manner - e.g. as an adjacency matrix with a single edge attribute.

This is probably not a satisfactory answer, but I really do think it depends on the graph(s) in question. Hope this is helpful in some way!

**Comment by ericmjl** [comment]
Agreed with @rossbar. That said, while there may not be a native way to store NetworkX graphs directly in Postgres databases, you can mimic a graph database by having two tables - one for nodes and their metadata, and one for edges and their metadata. 

While this page doesn’t directly deal with storing graph data in a database, it does give a hint on how to convert NetworkX graphs into two pandas data frames, which is the precursor to storage in a database.

**Reply by MikeB2019x** [reply]
"While this page ..."  did you mean this discussion or is there a link missing?

**Reply by ericmjl** [reply]
Ah, I'm so sorry, I forgot to paste the link! https://ericmjl.github.io/Network-Analysis-Made-Simple/03-practical/01-io/

**Comment by MikeB2019x** [comment, marked as answer]
Tnx!  My solution was to create 3 tables: meta, nodes, and edges.  Attributes are stored as json blobs.  It's not particularly elegant but it does allow me to store the structure and to do some very kludgy searching etc.  A cool thing (for me) was that a user can download a subgraph by specifying node and/or edge types ... that's actually pretty handy on a large graph.

**Reply by MridulS** [reply]
That sounds nice! @MikeB2019x
